src/palimpzest/query/operators/retrieve.py

```diff
@@ -92,19 +164,96 @@ def __call__(self, candidate: DataRecord) -> DataRecordSet:
+        start_time = time.time()
+
+        # check that query is a string or list of strings, otherwise return output with self.output_attr as None
+        query = getattr(candidate, self.search_attr)
+        query_is_str = isinstance(query, str)
+        query_is_list_of_str = isinstance(query, list) and all(isinstance(q, str) for q in query)
+        if not query_is_str and not query_is_list_of_str:
+            return self._create_record_set(
+                candidate=candidate,
+                top_k_results=None,
+                generation_stats=GenerationStats(),
+                total_time=time.time() - start_time,
+            )
+
+        # if query is a string, convert it to a list of strings
+        if query_is_str:
+            query = [query]
+
+        # compute embedding(s) if the index is a chromadb collection
+        gen_stats = GenerationStats()
+        if isinstance(self.index, Collection):
+            uses_openai_embedding_fcn = isinstance(self.index._embedding_function, OpenAIEmbeddingFunction)
+            assert uses_openai_embedding_fcn, "ChromaDB index must use OpenAI embedding function; see: https://docs.trychroma.com/integrations/embedding-models/openai"
+
+            model_name = self.index._embedding_function._model_name
+            err_msg = f"For Chromadb, we currently only support `text-embedding-3-small`; your index uses: {model_name}"
+            assert model_name == Model.TEXT_EMBEDDING_3_SMALL.value, err_msg
+
+            # compute embeddings
+            client = OpenAI()
+            embed_start_time = time.time()
+            response = client.embeddings.create(input=query, model=model_name)
+            embed_total_time = time.time() - embed_start_time
 
-        return record_set
+            # extract embedding(s)
+            query = [item.embedding for item in response.data]
+
+            # compute cost of embedding(s)
+            model_card = MODEL_CARDS[model_name]
+            total_input_tokens = response.usage.total_tokens
+            total_input_cost = model_card["usd_per_input_token"] * total_input_tokens
+            gen_stats = GenerationStats(
+                model_name=model_name,
+                total_input_tokens=total_input_tokens,
+                total_output_tokens=0.0,
+                total_input_cost=total_input_cost,
+                total_output_cost=0.0,
+                cost_per_record=total_input_cost,
+                llm_call_duration_secs=embed_total_time,
+            )
+
+        try:
+            top_results = self.search_func(self.index, query, self.k)
+        except Exception:
+            top_results = ["error-in-retrieve"]
+            os.makedirs("retrieve-errors", exist_ok=True)
+            ts = time.time()
+            with open(f"retrieve-errors/error-{ts}.txt", "w") as f:
+                f.write(str(query))
+
+        # filter top_results for the top_k_results
+        top_k_results = []
```

Comment:
Is this for the case the user-provided search_func does not return <= k entries? Should we be overriding the user in that case?